### ISSUE TITLE
cloudwatch - Add Alarms on Metric Math Expressions

### ISSUE THREAD

**Issue by nirpod** [terraform-providers/terraform-provider-aws#6551]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

need to add support for this new feature:
Amazon CloudWatch Launches Ability to Add Alarms on Metric Math Expressions

### New or Affected Resource(s)

aws cloudwatch alarms

### References

https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-cloudwatch-launches-ability-to-add-alarms-on-metric-math-expressions/

**Comment by nirpod** [comment]
+1

**Comment by dpetzold** [comment]
I looked at what a possible implementation would look like. My notes are below. Please let me know what you think.

The` aws_cloudwatch_metric_alarm` resource would be need to support a new nested block `metric` which would look like the following:

```
metric {                                                                                                             
    stat {                                                                                                           
        name = "ConnectionsFailed"                                                                                   
        namespace = "MyService"                                                                                      
        period = 60                                                                                                  
        stat = "Sum"                                                                                                 
    }                                                                                                                
    id = "m1"                                                                                                        
}                                                                                                                    
                                                                                                                     
metric {                                                                                                             
    stat {                                                                                                           
        name = "ConnectionsDropped"                                                                                  
        namespace = "MyService"                                                                                      
        period = 60                                                                                                  
        stat = "Sum"                                                                                                 
    }                                                                                                                
    id = "m2"                                                                                                        
}                                                                                                                    
                                                                                                                     
metric {                                                                                                             
    stat {                                                                                                           
        name = "RequestsThrottled"                                                                                   
        namespace = "MyService"                                                                                      
        period = 60                                                                                                  
        stat = "Sum"                                                                                                 
    }                                                                                                                
    id = "m3"                                                                                                        
}                                                                                                                     
                                                                                                                     
metric {                                                                                                             
    stat {                                                                                                           
        name = "ConnectionAttempts"                                                                                  
        namespace = "MyService"                                                                                      
        period = 60                                                                                                  
        stat = "Sum"                                                                                                 
    }                                                                                                                
    id = "m4"                                                                                                        
 }                                                                                                             
                                                                                                                     
metric {                                                                                                             
    id = "error_total"                                                                                               
    expression = "m1+m2+m3"                                                                                          
}                                                                                                                    
                                                                                                                     
metric {                                                                                                             
    id = "error_rate"                                                                                                
    expression = "(error_total/m4)*100"                                                                              
    label = "Total Connection Error Rate"
    return_data = "true"                                                                        
}
```
Where:
* `return_data` defaults to false.
* `stat` and `expression` are mutually exclusive.

Notice I folded in the `Metric` structure into `MetricStat` and renamed `MetricStat` to `stat`. I think that makes more sense from a configuration usage perspective but now it is no longer one-to-one. 

https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricAlarm.html
https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#CloudWatch.PutMetricAlarm
https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#MetricDataQuery
https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#MetricStat

**Comment by kakmt** [comment]
How's https://github.com/terraform-providers/terraform-provider-aws/pull/6833 going?

**Comment by benlamonica** [comment]
Hi @bflad,
Would you consider merging Metric Math Expression support in? Are there any specific concerns with this implementation? I need this functionality and have been waiting for a while now. Will you give us an idea of when this can be merged?

**Comment by shaiguitar** [comment]
Any update on this PR? Would be great to have a new version with working cw math expressions in place.

**Comment by shaiguitar** [comment]
Bump - looking forward to seeing this get in. If there's something else pending on this I am happy to take a look/help out. Thanks in advance!

**Comment by bflad** [comment]
Support for math expression metric alarms (via a new `metric_query` configuration block) has been merged and will release with version 1.59.0 of the Terraform AWS Provider, likely middle of this week. The resource documentation will include an example of how to set these up. 🚀

**Comment by shaiguitar** [comment]
Thanks @bflad - out of curiosity for the impatient; is there a method or docs on how to use a work-in-progress terraform AWS provider/git branch to test out the functionality until an official provider is released? 

Thanks for all the work you do.

**Comment by bflad** [comment]
@shaiguitar the [README instructions](https://github.com/terraform-providers/terraform-provider-aws/blob/master/README.md) should get you on your way to a custom built provider locally (e.g., a Go 1.11.4+ environment using `make build`) and points to the Terraform instructions for [installing a plugin](https://www.terraform.io/docs/plugins/basics.html#installing-plugins) since you will want to use a local custom built binary instead of one downloaded by `terraform init`. The only very recent change to those instructions as written is that working inside the `GOPATH` is optional now since we have migrated this codebase Go modules.

**Comment by shaiguitar** [comment]
Any update on a release with this feature (end of week)?

**Comment by bflad** [comment]
This has been released in [version 1.59.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1590-february-14-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by shaiguitar** [comment]
Verified this works as expected, thanks!
